src/soft.rs

```diff
@@ -393,6 +436,107 @@ impl HsmIdentity for SoftHsm {
         })
     }
+
+    fn identity_key_certificate_request(
+        &mut self,
+        mk: &Self::MachineKey,
+        loadable_key: &Self::LoadableIdentityKey,
+        cn: &str,
+    ) -> Result<Vec<u8>, HsmError> {
+        let id_key = self.identity_key_load(mk, loadable_key)?;
+
+        let mut req_builder = X509ReqBuilder::new().map_err(|ossl_err| {
+            error!(?ossl_err);
+            HsmError::X509RequestBuilder
+        })?;
+
+        let mut x509_name = X509NameBuilder::new().map_err(|ossl_err| {
+            error!(?ossl_err);
+            HsmError::X509NameBuilder
+        })?;
+
+        x509_name
+            .append_entry_by_text("CN", cn)
+            .map_err(|ossl_err| {
+                error!(?ossl_err);
+                HsmError::X509NameAppend
+            })?;
+
+        let x509_name = x509_name.build();
+        req_builder
+            .set_subject_name(&x509_name)
+            .map_err(|ossl_err| {
+                error!(?ossl_err);
+                HsmError::X509RequestSubjectName
+            })?;
+
+        match id_key {
+            SoftIdentityKey::Ecdsa256 { pkey, x509: _ }
+            | SoftIdentityKey::Rsa2048 { pkey, x509: _ } => {
+                req_builder.set_pubkey(&pkey).map_err(|ossl_err| {
+                    error!(?ossl_err);
+                    HsmError::X509RequestSetPublic
+                })?;
+
+                req_builder
+                    .sign(&pkey, MessageDigest::sha256())
+                    .map_err(|ossl_err| {
+                        error!(?ossl_err);
+                        HsmError::X509RequestSign
+                    })?;
+            }
+        }
+
+        req_builder.build().to_der().map_err(|ossl_err| {
+            error!(?ossl_err);
+            HsmError::X509RequestToDer
+        })
+    }
+
+    fn identity_key_associate_certificate(
+        &mut self,
+        mk: &Self::MachineKey,
+        loadable_key: &Self::LoadableIdentityKey,
+        certificate_der: &[u8],
+    ) -> Result<Self::LoadableIdentityKey, HsmError> {
+        let id_key = self.identity_key_load(mk, loadable_key)?;
+
+        // Verify the certificate matches our key
+        let certificate = X509::from_der(certificate_der).map_err(|ossl_err| {
+            error!(?ossl_err);
+            HsmError::X509FromDer
+        })?;
+
+        let certificate_pkey = certificate.public_key().map_err(|ossl_err| {
+            error!(?ossl_err);
+            HsmError::X509PublicKey
+        })?;
+
+        match id_key {
+            SoftIdentityKey::Ecdsa256 { pkey, x509: _ }
+            | SoftIdentityKey::Rsa2048 { pkey, x509: _ } => {
+                if !pkey.public_eq(&certificate_pkey) {
+                    return Err(HsmError::X509KeyMismatch);
+                }
+            }
+        };
+
+        // At this point we know the cert belongs to this key, so lets
+        // get it bound.
+
+        let mut cloned_key = loadable_key.clone();
```

Comment:
The alternative here is we can encrypt the x509 der to prevent tampering while it's not loaded. But we currently just check that it's public key matches on load, so that is likely enough. Thoughts?

Reply:
How would it get encrypted, where would its key come from etc?

Reply:
The way this works is there is a key hierarchy: Each key encrypts the children below it. 

```
┌────────────────────────┐
│  TPM 1                 │
│                        │
│   ┌───────────────┐    │
│   │   Primary 1   │    │
│   └───────────────┘    │
│           │            │
│           ▼            │
│   ┌───────────────┐    │
│   │  Storage Key  │    │
│   └───────────────┘    │
│           │            │
│           ▼            │
│    ┌────────────────┐  │
│   ┌┴───────────────┐│  │
│  ┌┴──────────────┐ ├┘  │
│  │  Child Keys   ├─┘   │
│  └───────────────┘     │
└────────────────────────┘
```

So here we have the authValue (some password/secret) that is used to decrypt the machine key, which itself then can be used to decrypt/encrypt the ID keys. 

At this point because we take the machine key (which is the parent to the id key) and we use it to load the id key, we have the material available to encrypt the x509 cert too. 

The question isn't about "where would we encrypt" or "where is it's key" but "why would we need to encrypt the x509 cert associated to a private key at all?". The private key is encrypted already when needed, so is there a need to encrypt the x509 part? Do we want to limit exposure of the content of that cert? Prevent alteration of the x509 (which we already defend from when we load the key).

Reply:
Thanks for the clarification. Certs are public. The private bit's denoted by its name, so ... my answer to the parent question would be "nope".